test/unit-luatest/vconnect_test.lua

```diff
@@ -0,0 +1,181 @@
+local timeout_opts = {timeout = vtest.wait_timeout}
+
+local test_group = t.group('vconnect')
+
+local cfg_template = {
+    sharding = {
+        replicaset = {
+            replicas = {
+                replica = {
+                    master = true,
+                },
+            },
+        },
+    },
+    bucket_count = 20,
+    test_user_grant_range = 'super',
+    replication_timeout = 0.1,
+    identification_mode = 'name_as_key',
+}
+local global_cfg
+
+test_group.before_all(function(g)
+    t.run_only_if(vutil.feature.persistent_names)
+    global_cfg = vtest.config_new(cfg_template)
+    vtest.cluster_new(g, global_cfg)
+    vtest.cluster_bootstrap(g, global_cfg)
+    vtest.cluster_wait_vclock_all(g)
+end)
+
+test_group.after_all(function(g)
+    g.cluster:stop()
+end)
+
+--
+-- Test, that conn_vconnect_wait fails to get correct
+-- result. Connection should be closed.
+--
+test_group.test_vconnect_no_result = function(g)
+    local _, rs = next(vreplicaset.buildall(global_cfg))
+    g.replica:exec(function()
+        rawset(_G, '_call', ivshard.storage._call)
+        ivshard.storage._call = nil
+    end)
+
+    -- Drop connection in order to make replicaset to recreate it.
+    rs.master.conn = nil
+    local ret, err = rs:callrw('get_uuid', {}, timeout_opts)
+    t.assert_str_contains(err.message, "_call' is not defined")
+    t.assert_equals(ret, nil)
+    -- Critical error, connection should be closed.
+    t.assert_equals(rs.master.conn.state, 'closed')
+
+    g.replica:exec(function()
+        ivshard.storage._call = _G._call
+    end)
+end
+
+--
+-- Test, that conn_vconnect_wait fails, when future is nil.
+--
+test_group.test_vconnect_no_future = function(g)
+    local _, rs = next(vreplicaset.buildall(global_cfg))
+    g.replica:exec(function()
+        rawset(_G, '_call', ivshard.storage._call)
+        -- Future should not appear at all.
+        ivshard.storage._call = function()
+            while true do
+                ifiber.sleep(0.1)
+            end
+        end
+    end)
+
+    rs.master.conn = nil
+    local ret, err = rs:callrw('get_uuid', {}, small_timeout_opts)
+    t.assert(verror.is_timeout(err))
+    t.assert_equals(ret, nil)
+    t.assert_not_equals(rs.master.conn.state, 'closed')
+
+    g.replica:exec(function()
+        ivshard.storage._call = _G._call
+    end)
+end
+
+--
+-- Test, that conn_vconnect_check fails, when future's result is nil.
+--
+test_group.test_vconnect_check_no_future = function(g)
+    local _, rs = next(vreplicaset.buildall(global_cfg))
+    g.replica:exec(function()
+        rawset(_G, '_call', ivshard.storage._call)
+        ivshard.storage._call = nil
+    end)
+
+    rs.master.conn = nil
+    local opts = table.deepcopy(timeout_opts)
+    opts.is_async = true
+    local ret, err = rs:callrw('get_uuid', {}, timeout_opts)
+    t.assert_str_contains(err.message, "_call' is not defined")
+    t.assert_equals(ret, nil)
+    t.assert_equals(rs.master.conn.state, 'closed')
+
```

Comment:
You set `opts` but never use it. This test is identical to `test_vconnect_no_result`. Something is wrong here.

Reply:
Thank you for noticing! We must use `opts` and retry until the time, actual error is shown. Fixed